### ISSUE TITLE
Context in /api/generate response grows too big.

### ISSUE THREAD

**Issue by slouffka** [ollama/ollama#5980]
### What is the issue?

I'm coding my own Chat UI for Ollama and using context feature to implement dialog mode. So every time Ollama generates a response the returned context (embeddings) is saved into chat object. On the next prompt this context is passed into `/api/generate` then after response resulting context is saved into chat object again.

After upgrading to latest Ollama I've noticed generation speed degraded 
considerably and the context returned by `/api/generate` grows too fast compared to previous versions.

Looks like it doubles context size after each generation and soon in relatively small chat with 26 messages it becomes like 3-7Mb in size which causes my UI being unresponsive and also browser freezes because it has to process such a huge amount of data (mostly for debugging like converting JSON to string, but this is not normal anyway). When earlier (at least for the 0.2.1 version I've used) it could be around 8-16Kb which is totally fine and also fits model capacity.

This is pretty hard to measure (and I don't know how to) but I've also notices that with latest Ollama newer models like gemma2 or llama3.1 do not adhere to context as well as some older models like mistral on earlier Ollama version. This could be related to a context changes, which was broken since 0.2.2 then response was fixed but it looks like the fix was not completely correct.

### OS

Linux

### GPU

Nvidia

### CPU

Intel

### Ollama version

0.3.0

**Comment by wangzhezhe** [comment]
Hi, I have similar issues and not sure how to solve it properly. According to description about the `context` [here](https://github.com/ollama/ollama/blob/main/docs/api.md),  the `context` returned by current conversation can be used as input for the next conversation. One issue of using context by this way is that the `context` could expand quickly. If I only want to track results of last 2-3 questions as the context (instead of all history), what are proper ways to do that, thanks!

**Comment by slouffka** [comment]
There was no real issue in using context this way cause it's actually smaller and more efficient than sending whole chat history in JSON. Because this way context is already stored as embeddings. To me it also feels more efficient to process on Ollama side because it does not need to convert text to embeddings but using them as is to provide context for an LLM.

I've written the client code for this feature and using my Chat UI for a 6+ month without any changes. It was perfectly working and clean solution but now it's broken so I'm stick with ollama version 0.2.1 or I have an option to switch to an Open AI like chat endpoint where you pass chat history as a context. Personally I don't feel it's a lot worse but I would not like to rewrite working app because of a bug. Though last broken version I've tried was 0.3.0, maybe there was a fix I dunno. But this issue is still unsolved.

**Comment by wangzhezhe** [comment]
Yeah, it seems that I should use api/chat instead of api/generate, thanks for the information！@slouffka

**Comment by mdhuzaifapatel** [comment]
> There was no real issue in using context this way cause it's actually smaller and more efficient than sending whole chat history in JSON. Because this way context is already stored as embeddings. To me it also feels more efficient to process on Ollama side because it does not need to convert text to embeddings but using them as is to provide context for an LLM.
> 
> I've written the client code for this feature and using my Chat UI for a 6+ month without any changes. It was perfectly working and clean solution but now it's broken so I'm stick with ollama version 0.2.1 or I have an option to switch to an Open AI like chat endpoint where you pass chat history as a context. Personally I don't feel it's a lot worse but I would not like to rewrite working app because of a bug. Though last broken version I've tried was 0.3.0, maybe there was a fix I dunno. But this issue is still unsolved.

Hi can you please tell me how to use context of previous responses and sent it as context to current one. 
I'm unable to achieve it. Everytime I'm hitting Ollama API it's looks like it is considering it as a new session. I'm not sure, can you please let me know how to implement it. 
Pls provide code example if possible as you have already worked on it. Thanks in advance

**Comment by mdhuzaifapatel** [comment]
Thanks!
